virtual_ship/sailship.py

```diff
@@ -229,61 +151,55 @@ def sailship(config: VirtualShipConfig):
-    for startloc, endloc in zip(config.route_coordinates, config.route_coordinates[1:]):
-        # iterate over each coordinate and the next coordinate
-        # last coordinate has no next coordinate and is skipped
-
-        # get locations between start and end, seperate by 5 minutes of cruising
-        # excludes final point, but this is added explicitly after this loop
-        int_points = geod.inv_intermediate(
-            startloc.lon,
-            startloc.lat,
-            endloc.lon,
-            endloc.lat,
-            del_s=CRUISE_SPEED * dt.total_seconds(),
-            initial_idx=0,
-            return_back_azimuth=False,
+    # print("Postprocessing..")
+    # postprocess()
+
+    # print("All data has been gathered and postprocessed, returning home.")
+
+    # time_past = time - config.waypoints[0].time
+    # cost = costs(config, time_past)
+    # print(f"This cruise took {time_past} and would have cost {cost:,.0f} euros.")
+
+
+def _verify_waypoints(
+    waypoints: list[Waypoint], ship_speed: float, projection: pyproj.Geod
+) -> None:
+    # check first waypoint has a time
+    if waypoints[0].time is None:
+        raise PlanningError("First waypoint must have a specified time.")
+
+    # check waypoint times are in ascending order
+    timed_waypoints = [wp for wp in waypoints if wp.time is not None]
+    if not all(
+        [
+            next.time >= cur.time
+            for cur, next in zip(timed_waypoints, timed_waypoints[1:])
+        ]
+    ):
+        raise PlanningError(
+            "Each waypoint should be timed after all previous waypoints"
         )
 
-        sample_points.extend(
-            [
-                Location(latitude=lat, longitude=lon)
-                for lat, lon in zip(int_points.lats, int_points.lons, strict=True)
-            ]
-        )
+    # check that ship will arrive on time at each waypoint (in case nothing goes wrong)
+    time = waypoints[0].time
+    for wp, wp_next in zip(waypoints, waypoints[1:]):
+        match wp.instrument:
+            case InstrumentType.CTD:
+                time += timedelta(minutes=20)
 
-    # explitly include final point which is not added by the previous loop
-    sample_points.append(
-        Location(
-            latitude=config.route_coordinates[-1].lat,
-            longitude=config.route_coordinates[-1].lon,
+        geodinv: tuple[float, float, float] = projection.inv(
+            wp.location.lon, wp.location.lat, wp_next.location.lon, wp_next.location.lat
         )
-    )
+        distance = geodinv[2]
 
-    return sample_points
+        time_to_reach = timedelta(seconds=distance / ship_speed)
+        arrival_time = time + time_to_reach
+
+        if wp_next.time is None:
+            time = arrival_time
+        elif arrival_time > wp_next.time:
+            raise PlanningError(
+                "Waypoint planning is not valid: would arrive too late a waypoint."
+            )
```

Comment:
Give a bit more context? Which point would be arrived too late? This helps debugging schedules later